enhancements/installer/pre-flight-validations.md

```diff
@@ -0,0 +1,194 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+- Will the validations be run automatically on `cluster create` or will it be
+  an explicit action?
+- If explicit action, how will the validations be enabled? E.g. adding
+  a `--dry-run` option to the installer vs. a separate subcommand or even
+  a separate binary.
+- When enabling the validations, will it also perform the installation or
+  simply run the validations?
+- Can the installer override validations that failed?
+- How are failures and warnings reported to the user?
+
+## Summary
+
+One of the guiding principles of OCP 4 is that the installation should always
+succeed. This goal is relatively easy to implement for public cloud platforms
+where we can make assumptions about services being available, or performance
+meeting requirements, however this is not the case with private clouds where
+each cloud is unique.
+
+It is currently not possible to tell with confidence if an installation will be
+successful or not for the following platforms:
+- BareMetal
+- OpenStack
+- oVirt
+- vSphere
+
+For this purpose, we propose to implement a framework allowing the installer to
+run pre-flight validations in order to certify that all pre-requisites are met
+to successfully install OpenShift in the selected environment.
+
+## Motivation
+
+This section is for explicitly listing the motivation, goals and non-goals of
+this proposal. Describe why the change is important and the benefits to users.
+
+### Goals
+
+By having an automated way to identify potential issues early, before the
+deployment even started, we want to reduce wasted time and resources, customer
+escalations, and improve the perception of OpenShift deployments on-premise.
+
+The installer already performs some validations:
+- checks all required fields are set and that the data is in the right format
+- some basic validation, such as networks do not overlap
+
+However, this doesn't check that the environment is suitable to install OpenShift:
+- pull secret is valid to fetch the container images
+- the tenant has adequate quota and the flavors' specifications are within the
+  recommended ranges.
+- for user-provided networks, check the subnets have a DHCP server and valid
+  DNS to reach the cloud's enpoints
+- necessary cloud services are available
+- storage performance
+
+### Non-Goals
+
+Implementing every useful validation we can think of is out-of-scope:
+- The potential number of validations is huge
+- We expect more validations to be added over time as new issues are
+  discovered.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Stories
+
+#### On-premise deployments
+
+As an administrator of an OpenShift cluster, I would like to verify that my
+OpenStack cloud meets all the performance, service and networking requirements
+necessary for a successful deployment.
+
```

Comment:
As noted below, we want to spin a VM mimicking a `master` node. To me this means that either implicitly or explicitly, we are validating an `install-config` as well.

To be even clearer about this point, I'd add a reference to the aforementioned OpenShift admin wanting to check that his cluster's configuration is valid and appropriate for the target infrastructure.

Reply:
I guess this is ok for OpenStack, but for bare metal it would be almost like if a complete installation is executed (boot one of the bare metal masters/workers with a 'special image' that can execute the prechecks, run the checks, report back & tear down the baremetal... including how to manage the pxe, etc. that is required to boot the bare metal host)